File: parameterized-labs/chap01/PokerLab/starter-code/equivalent-java/Card.java
/*
 * PokerLab assignment and starter code (c) 2018-23 Christopher A. Bohn
 */

/* ************************************************************************//**
 *
 * Methods to populate a card's fields and to create a printable
 * representation of a card.
 *
 * @author Christopher A. Bohn
 *
 ******************************************************************************/
public class Card {

    public static final int MINIMUM_FACE_VALUE = 2;
    public static final int MAXIMUM_FACE_VALUE = 10;
    public static final int MINIMUM_VALUE = MINIMUM_FACE_VALUE - 1;
    public static final int MAXIMUM_VALUE = MAXIMUM_FACE_VALUE + 3;
    public static final int NUMBER_OF_CARDS = 52;

    public enum Suit {CLUBS, DIAMONDS, HEARTS, SPADES}

    public int value;
    public Suit suit;

    /**
     * Initializes a Card.
     *
     * @param value the new Card's value
     * @param suit  the new Card's suit
     */
    public Card(int value, Suit suit) {
        this.value = value;
        this.suit = suit;
    }

    /**
     * Returns the printable representation of a Card.
     *
     * @return the printable representation of this Card
     */
    public String toString() {
        String valueString;
        String suitsString = suit.toString();
        if (value < MINIMUM_VALUE || value > MAXIMUM_VALUE) {                           // Illegal values
            valueString = "UNKNOWN";
        } else if ((value >= MINIMUM_FACE_VALUE) && (value <= MAXIMUM_FACE_VALUE)) {    // Number card
            valueString = Integer.toString(value);
        } else {                                                                        // Ace or face card
            switch (value) {
                case MINIMUM_VALUE:
                    valueString = "ACE";
                    break;
                case MAXIMUM_VALUE - 2:
                    valueString = "JACK";
                    break;
                case MAXIMUM_VALUE - 1:
                    valueString = "QUEEN";
                    break;
                case MAXIMUM_VALUE:
                    valueString = "KING";
                    break;
                default:
                    valueString = "DEADCODE";       // This line is unreachable
            }
        }
        return valueString + " of " + suitsString;
    }
}

File: lab-templates/PokerLab/src/card.c
/**************************************************************************//**
 *
 * @file card.c
 *
 * @author (STUDENT -- TYPE YOUR NAME HERE)
 *
 * @brief Functions to populate a card's fields and to create a printable
 *      representation of a card.
 *
 ******************************************************************************/

/*
 * PokerLab assignment and starter code (c) 2018-26 Christopher A. Bohn
 * PokerLab solution (c) the above-named student
 */

#include <stdio.h>
#include "card.h"


/**
 * Initializes a card_t.
 *
 * @param value the new card_t's value
 * @param suit  the new card_t's suit
 * @param destination a pointer to the card_t whose fields will be populated
 * @return the `destination` pointer
 */
card_t *create_card(int value, suit_t suit, card_t *destination) {
    /* ADD CODE HERE TO ASSIGN VALUE AND SUIT TO THE DESTINATION'S FIELDS */

    return destination;
}

/**
 * Returns the printable representation of a card_t.
 *
 * @pre The destination must have at least 21 bytes allocated.
 *
 * @param card a pointer to the card_t to be represented as a string
 * @param destination a pointer to a buffer with at least 21 bytes allocated for
 *      the string
 * @return the `destination` pointer
 */
char *card_to_string(card_t const *card, char *destination) {
    /* ASSIGN `CARD`'S VALUE TO THE `VALUE` VARIABLE */
    int value = -1;
    /* ASSIGN `CARD`'S SUIT TO THE `SUIT` VARIABLE */
    suit_t suit = -1;
    char *value_string;
    char *suit_string;
    if (suit >= 0 && suit < NUMBER_OF_SUITS) {
        suit_string = suit_to_string[suit];
    } else {
        suit_string = "UNKNOWN";
    }
    char buffer[3];
    if ((value < MINIMUM_VALUE) || (value > MAXIMUM_VALUE)) {                       // Illegal values
        value_string = "UNKNOWN";
    } else if ((value >= MINIMUM_NUMBER_VALUE) && (value <= MAXIMUM_NUMBER_VALUE)) {    // Number card
        value_string = buffer;
        /* THE SECOND ARGUMENT NEEDS THE CONVERSION SPECIFIER THAT YOU WOULD USE TO PRINT AN INTEGER */
        sprintf(value_string, "", value);
    } else {                                                                        // Ace or face card
        switch (value) {
            case ACE:
                value_string = "ACE";
                break;
            case JACK:
                value_string = "JACK";
                break;
            case QUEEN:
                value_string = "QUEEN";
                break;
            case KING:
                value_string = "KING";
                break;
            default:
                value_string = "DEADCODE";      // This line is unreachable
        }
    }
    sprintf(destination, "%s of %s", value_string, suit_string);
    return destination;
}
